### PR TITLE
Unsubscribe from notifications

### PR DESCRIPTION
**Pull request by aapeliv**
**Backend checklist**
- [ ] Formatted my code by running `autoflake -r -i --remove-all-unused-imports src && isort . && black .` in `app/backend`
- [ ] Added tests for any new code or added a regression test if fixing a bug
- [ ] All tests pass
- [ ] Run the backend locally and it works
- [ ] Added migrations if there are any database changes, rebased onto `develop` if necessary for linear migration history

### CODE REVIEW

File: app/backend/src/couchers/notifications/unsubscribe.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+from proto.internal import unsubscribe_pb2
+
+logger = logging.getLogger(__name__)
+
+
+def _generate_unsubscribe_link(payload):
+    msg = payload.SerializeToString()
+    sig = generate_hash_signature(message=msg, key=UNSUBSCRIBE_KEY)
+    return urls.unsubscribe_link(
+        payload=urlsafe_b64encode(msg).decode("utf8"), sig=urlsafe_b64encode(sig).decode("utf8")
```
Comment (kalvdans) [inline app/backend/src/couchers/notifications/unsubscribe.py]:
```suggestion
        payload=urlsafe_b64encode(msg).decode("ascii"), sig=urlsafe_b64encode(sig).decode("ascii")
```

File: app/backend/src/couchers/notifications/unsubscribe.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+            user_id=notification.user_id,
+            topic_action=unsubscribe_pb2.UnsubscribeTopicAction(
+                topic=notification.topic,
+                action=notification.action,
+            ),
+        )
+    )
+
+
+def unsubscribe(request, context):
```
Comment (kalvdans) [inline app/backend/src/couchers/notifications/unsubscribe.py:60]: Document the return value.

File: app/proto/auth.proto [hunk 3]
```diff
@@ -243,3 +247,12 @@ message ContributorForm {
   repeated string contribute_ways = 5;
   string expertise = 6;
 }
+
+message UnsubscribeReq {
+  bytes payload = 1;
+  bytes sig = 2;
```
Comment (kalvdans) [inline app/proto/auth.proto:254]: Please describe if these should be base64url encoded or raw bytes.

File: app/proto/auth.proto [hunk 4]
```diff
@@ -243,3 +247,12 @@ message ContributorForm {
...
+}
+
+message UnsubscribeRes {
+  string response = 1;
```
Comment (kalvdans) [inline app/proto/auth.proto]: Please describe that actually getting an UnsubscribeRes always means that the operation was successful, regardless of the string.

**Review by kalvdans — COMMENTED** [review]
Backend looks good!

It seems protobuf serialization is not deterministic, but since the payload is never unpacked+repacked by the client it will work :)

File: app/web/features/auth/notifications/Unsubscribe.tsx [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+  const {
+    data,
+    error,
+    isLoading,
+    isSuccess,
+    mutate: unsubscribe,
+  } = useMutation<UnsubscribeRes.AsObject, RpcError, UnsubscribeParams>(
+    async ({ payload, sig }) => {
+      return await service.auth.unsubscribe(payload, sig);
+    }
```
Comment (lucaslcode) [inline app/web/features/auth/notifications/Unsubscribe.tsx]:
```suggestion
    async ({ payload, sig }) => service.auth.unsubscribe(payload, sig)
```

File: app/web/features/auth/notifications/Unsubscribe.tsx [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+            <Alert severity="success">{data?.response || t("global:success")}</Alert>
+          ) : error ? (
+              <Alert severity="error">
+                {t("auth:unsubscribe.error_message", {
+                  message: error.message,
+                })}
+              </Alert>
+            ): <Button onClick={() => unsubscribe({payload, sig})}>{t("auth:unsubscribe.button_text")}</Button>
+          )
+        }
```
Comment (lucaslcode) [inline app/web/features/auth/notifications/Unsubscribe.tsx]:
```suggestion
    {error && (
      <Alert severity="error">
        {t("auth:unsubscribe.error_message", {
          message: error.message,
        })}
      </Alert>
    )}
    {isSuccess && <Alert severity="success">{data?.response || t("global:success")}</Alert>}
    <Button
      onClick={() => unsubscribe({payload, sig})}
      loading={isLoading}
    >
      {t("auth:unsubscribe.button_text")}
    </Button>
```

File: app/web/pages/unsubscribe.tsx [hunk 7]
```diff
@@ -0,0 +1,21 @@
+import { appGetLayout } from "components/AppRoute";
+import Unsubscribe from "features/auth/notifications/Unsubscribe";
+import { GetStaticProps } from "next";
+import nextI18nextConfig from "next-i18next.config";
+import { serverSideTranslations } from "next-i18next/serverSideTranslations";
+
+export const getStaticProps: GetStaticProps = async ({ locale }) => ({
```
Comment (lucaslcode) [inline app/web/pages/unsubscribe.tsx]: There's a helper function for this in `i18n/server-side-translations.ts`: `translationStaticProps`

File: app/web/pages/unsubscribe.tsx [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+  props: {
+    ...(await serverSideTranslations(
+      locale ?? "en",
+      ["global", "auth"],
```
Comment (lucaslcode) [inline app/web/pages/unsubscribe.tsx]:
```suggestion
      [GLOBAL, AUTH],
```

File: app/web/pages/unsubscribe.tsx [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+      nextI18nextConfig
+    )),
+  },
+});
+
+export default function ConfirmEmailPage() {
```
Comment (lucaslcode) [inline app/web/pages/unsubscribe.tsx]:
```suggestion
export default function UnsubscribePage() {
```

File: app/web/pages/unsubscribe.tsx [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+  return <Unsubscribe />;
+}
+
+ConfirmEmailPage.getLayout = appGetLayout({ isPrivate: false });
```
Comment (lucaslcode) [inline app/web/pages/unsubscribe.tsx]:
```suggestion
UnsubscribePage.getLayout = appGetLayout({ isPrivate: false });
```

**Review by lucaslcode — CHANGES_REQUESTED** [review]

File: app/proto/auth.proto (hunk 3 above)
Comment (aapeliv) [reply app/proto/auth.proto:254]: base64 would be `string` instead of `bytes`, but I've added a comment as well!

**Review by lucaslcode — APPROVED** [review]

**Pull request merged**